api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java

```diff
@@ -0,0 +1,113 @@
+import static org.junit.Assert.assertNotNull;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.Before;
+import org.junit.rules.ExpectedException;
+import org.openmrs.CodedOrFreeText;
+import org.openmrs.Condition;
+import org.openmrs.Concept;
+import org.openmrs.ConceptName;
+import org.openmrs.ConditionVerificationStatus;
+import org.openmrs.Diagnosis;
+import org.openmrs.Encounter;
+import org.openmrs.Patient;
+import org.openmrs.User;
+import org.openmrs.api.db.DiagnosisDAO;
+import org.openmrs.test.BaseContextSensitiveTest;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class HibernateDiagnosisDAOTest extends BaseContextSensitiveTest {
+
+	private static final String DIAGNOSIS_XML = "org/openmrs/api/include/HibernateDiagnosisDAOTestDataset.xml";
+
+	@Autowired
+	DiagnosisDAO diagnosisDAO;
+
+	@Before
+	public void setUp() {
+		executeDataSet(DIAGNOSIS_XML);
+
+	}
+
+	@Test
+	public void shouldSaveDiagnosis() {
+		CodedOrFreeText codedOrFreeText = new CodedOrFreeText(new Concept(4),
+			new ConceptName(5089), "non coded");
+		int diagnosisId = 5;
+		
+		Diagnosis diagnosis = new Diagnosis();
+		diagnosis.setEncounter(new Encounter(3));
+		diagnosis.setRank(2);
+		diagnosis.setId(diagnosisId);
+		diagnosis.setCertainty(ConditionVerificationStatus.CONFIRMED);
+		diagnosis.setPatient(new Patient(2));
+		diagnosis.setCondition(new Condition());
+		diagnosis.setDiagnosis(codedOrFreeText);
+		diagnosis.setUuid("4e663d96-6b78-11e0-93c3-18a9b5e044dc");
+		diagnosis.setCreator(new User(1));
+		diagnosis.setVoided(false);
+		diagnosis.setDateCreated(new Date());
+
+		diagnosisDAO.saveDiagnosis(diagnosis);
+
+		Diagnosis savedDiagnosis = diagnosisDAO.getDiagnosisById(diagnosisId);
+		
+		assertEquals(diagnosis.getUuid(), savedDiagnosis.getUuid());
+		assertEquals(diagnosis.getVoided(), savedDiagnosis.getVoided());
+		assertEquals(diagnosis.getRank(), savedDiagnosis.getRank());
+		assertEquals(diagnosis.getCertainty(), savedDiagnosis.getCertainty());
+		assertEquals(diagnosis.getCreator(), savedDiagnosis.getCreator());
+		assertEquals(diagnosis.getCondition(), savedDiagnosis.getCondition());
+		assertEquals(diagnosis.getPatient(), savedDiagnosis.getPatient());
+		assertEquals(diagnosis.getEncounter(), savedDiagnosis.getEncounter());
+	}
+
+	@Test
+	public void shouldGetDiagnosisByUuid() {
+		Diagnosis diagnosis = diagnosisDAO.getDiagnosisByUuid("4e663d66-6b78-11e0-93c3-18a905e044dc");
+		assertEquals(1, (int) diagnosis.getId());
+	}
+
+	@Test
+	public void shouldGetDiagnosisById() {
+		Diagnosis diagnosis = diagnosisDAO.getDiagnosisById(1);
+		assertEquals("4e663d66-6b78-11e0-93c3-18a905e044dc", diagnosis.getUuid());
+	}
+
+	@Test
+	public void shouldGetActiveDiagnoses() {
+		assertEquals(2, diagnosisDAO.getActiveDiagnoses(new Patient(2)).size()); 
+	}
+
+	@Test
+	public void shouldGetDiagnoses() {
+		assertEquals(2, diagnosisDAO.getDiagnoses(new Encounter(3)).size());
```

Comment:
Since this test was passing before you fixed d.encounter.encounterId in the getDiagnoses(Encounter) method, is the test really properly written?

Reply:
It is. I tested it with bad data and it failed with the expected results